### PR TITLE
feat(mme): AMF Support for default slice type

### PR DESCRIPTION
**Pull request by panyogesh**
Fixed :
  1. Support added for default slice type (sST/SD) in mconfig.
  2. Support added for PLMN SUPPORT in config file having
     plmn to slice mapping in ngap response.
  3. Fixed te configuration read issue for amf name.
  4. Removed a global varaible buf_plmn.
  5. Add unit test infra for ngap setup request handler.

Test:
  1. Unit testing.
  2. Call flow testing with simulators.
 
Scenarios Covered :
1. No SliceType / No Slice Differentiator
2. With Slice Type as 1
3. With Mismatch Slice Types
4. With Matching SD Type
5. With Mismatching SD Type

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by panyogesh** [comment]
Ngap-Setup-Response-with-SD-Sep-15

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c [hunk 1]
```diff
@@ -500,159 +534,144 @@ status_code_e ngap_amf_handle_ng_setup_request(
...
+
+  /* Config Parameters */
+  Ngap_NGSetupResponse_t* ng_setup_response = NULL;
+
+  // AMF_NAME
+  Ngap_NGSetupResponseIEs_t* ie             = NULL;
+  Ngap_PLMNSupportList_t* plmn_support_list = NULL;
 
-  pdu.present = Ngap_NGAP_PDU_PR_successfulOutcome;
+  amf_config_read_lock(&amf_config);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c:547]: I don't see where you are unlocking this.

File: lte/protos/mconfig/mconfigs.proto [hunk 2]
```diff
@@ -356,6 +356,24 @@ message MME {
     SentryConfig sentry_config = 45;
     
     bool enable_converged_core = 46;
+
```
Comment (ardzoht) [inline lte/protos/mconfig/mconfigs.proto:359]: Thanks for updating these

File: lte/gateway/configs/gateway.mconfig [hunk 3]
```diff
@@ -39,7 +39,8 @@
       "ipv6DnsAddress": "2001:4860:4860:0:0:0:0:8888",
       "natEnabled": true,
       "congestionControlEnabled": true,
-      "enableConvergedCore": false
+      "enableConvergedCore": false,
+      "defaultSliceServiceType": "1",
```
Comment (ardzoht) [inline lte/gateway/configs/gateway.mconfig:43]: this should not have a trailing comma, as this is the last entry of the JSON object
```suggestion
      "defaultSliceServiceType": "1"
```

**Review by ardzoht — CHANGES_REQUESTED** [review]
let's remove the trailing comma on `defaultSliceServiceType` on the default gateway.mconfig, as this will break the JSON parsing

File: lte/gateway/configs/templates/mme.conf.template [hunk 4]
```diff
@@ -352,4 +353,8 @@ NGAP :
     # DNS address communicated to UEs
     DEFAULT_DNS_IPV4_ADDRESS     = "{{ ipv4_dns }}";
     DEFAULT_DNS_SEC_IPV4_ADDRESS = "{{ ipv4_sec_dns }}";
+    AMF_NAME = "{{ amf_name }}";
+    PLMN_SUPPORT_LIST = (
+       { MCC="{{ mcc }}" ; MNC="{{ mnc }}"; DEFAULT_SLICE_SERVICE_TYPE="{{ default_slice_service_type }}" ;                         DEFAULT_SLICE_DIFFERENTIATOR="{{ default_slice_differentiator }}"; }
```
Comment (ulaskozat) [inline lte/gateway/configs/templates/mme.conf.template]: whys is there so many blank spaces before `DEFAULT_SLICE_DIFFERENTIATOR`?

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c [hunk 5]
```diff
@@ -500,159 +534,145 @@ status_code_e ngap_amf_handle_ng_setup_request(
...
-  OCTET_STRING_fromBuf(&SliceItem->s_NSSAI.sST, (char*) &sst, 1);
+  amf_config_unlock(&amf_config);
+}
 
-  ASN_SEQUENCE_ADD(
-      &plmnItem->sliceSupportList.list,
-      SliceItem);  // adding slice item to slice list
+status_code_e ngap_generate_ng_setup_response(
+    ngap_state_t* state, gnb_description_t* gnb_association) {
+  int i;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c]: legit github comment about unused variable.

File: lte/gateway/c/core/oai/tasks/amf/amf_config.c [hunk 6]
```diff
@@ -662,7 +665,98 @@ int amf_config_parse_file(amf_config_t* config_pP) {
...
+          }
+
+          if ((config_setting_lookup_string(
+                  sub2setting, NGAP_CONFIG_PLMN_SUPPORT_SST, &set_sst)) &&
+              strlen(set_sst)) {
+            config_pP->plmn_support[i].s_nssai.sst = (uint8_t) atoi(set_sst);
+          } else {
+            config_pP->plmn_support[i].s_nssai.sst =
+                NGAP_S_NSSAI_ST_DEFAULT_VALUE;
+          }
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_config.c]: we should not do this here. Please add it into config initialization and here it can be overwritten if specified in the config.

Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/amf/amf_config.c]: +1

Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/amf/amf_config.c]: Comments addressed.

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c (hunk 5 above)
Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c]: Comments addressed.

File: lte/gateway/configs/templates/mme.conf.template (hunk 4 above)
Comment (panyogesh) [reply lte/gateway/configs/templates/mme.conf.template]: Comments addressed.

**Comment by panyogesh** [comment]
Default-NGAP-Setup-Request-Sep-16
Ngap-Setup-Request-with-SD-Sep-16
Ngap-Setup-Response-with-SD-Sep-16

**Comment by panyogesh** [comment]
[Test-Logs-Slice-SD-Support-Sep-16.txt](https://github.com/magma/magma/files/7177921/Test-Logs-Slice-SD-Support-Sep-16.txt)

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 7]
```diff
@@ -282,8 +282,7 @@ def _get_default_slice_differentiator_type_config(service_mconfig: object) -> st
 
     Returns: slice differentiator config value
     """
-
-    if service_mconfig.default_slice_differentiator is not None:
+    if len(service_mconfig.default_slice_differentiator):
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: what is the reason for this change?

**Comment by ulaskozat** [comment]
@rdefosse is Magma-OAI-Jenkins hanging?

**Review by ulaskozat — APPROVED** [review]
LGTM with a nit comment

**Review by ardzoht — APPROVED** [review]
lgtm

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 7 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: With GRPC/Protobuf an empty string is "" (len = 0)... Tested with following program to confirm                                
    if service_mconfig.default_slice_differentiator is not None:
        os.write(fd, str.encode (" ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -"))
        os.write(fd, str.encode(str(len(service_mconfig.default_slice_differentiator))))

Output : ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -0

Hence in mme.conf the default value "" and not the Default Value return by generate_oai_config script..

I guess we need to change other string variables on similar lines.

Comment (ulaskozat) [reply lte/gateway/python/scripts/generate_oai_config.py]: if these things are undefined on the orc8r side, I think the generated gateway.mconfig will not have the fields at all.  @ardzoht what do you say?

Comment (ardzoht) [reply lte/gateway/python/scripts/generate_oai_config.py]: The only case where you should expect an empty string from generated mconfigs is if the cloud config specifically returns that, but that should not be the case, as the default value for string types on the cloud configs is `"string"`. 

And if the config is not specified (as Ulas points out), there won't be any entry on the config at all. So `if service_mconfig.default_slice_differentiator is not None` is the correct handler for this. @panyogesh can you update it?

Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: My observation (DEFAULT VALUE : 0xffffff)

1. No entry for defaultSliceDifferentiator with : len(service_mconfig.default_slice_differentiator)   
-------------------------------------------------------------------------------------------------------------
vagrant@magma-dev-focal:~/magma/lte/gateway$ cat configs/gateway.mconfig | grep defaultSliceDifferentiator
vagrant@magma-dev-focal:~/magma/lte/gateway$

vagrant@magma-dev-focal:~/magma/lte/gateway$ cat ~/DUMP_SLICE
 ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -0vagrant@magma-dev-focal:~/magma/lte/gateway$

vagrant@magma-dev-focal:~/magma/lte/gateway$ cat /var/opt/magma/tmp/mme.conf  | grep  DEFAULT_SLICE_DIFFERENTIATOR
       { MCC="001" ; MNC="01"; DEFAULT_SLICE_SERVICE_TYPE="1" ; DEFAULT_SLICE_DIFFERENTIATOR="0xffffff"; }   <<<<<< Default value Set as expected.
vagrant@magma-dev-focal:~/magma/lte/gateway$

2. No entry for defaultSliceDifferentiator with : if service_mconfig.default_slice_differentiator is not None: 
--------------------------------------------------------------------------------------------------------------
vagrant@magma-dev-focal:~/magma/lte/gateway$  cat configs/gateway.mconfig | grep defaultSliceDifferentiator
vagrant@magma-dev-focal:~/magma/lte/gateway$

vagrant@magma-dev-focal:~/magma/lte/gateway$ cat ~/DUMP_SLICE
 ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -0vagrant@magma-dev-focal:~/magma/lte/gateway$

vagrant@magma-dev-focal:~/magma/lte/gateway$ cat /var/opt/magma/tmp/mme.conf  | grep  DEFAULT_SLICE_DIFFERENTIATOR
       { MCC="001" ; MNC="01"; DEFAULT_SLICE_SERVICE_TYPE="1" ; DEFAULT_SLICE_DIFFERENTIATOR=""; }
vagrant@magma-dev-focal:~/magma/lte/gateway$  <<<<<< Default value not set

Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: [Length-Experiment-Output.txt](https://github.com/magma/magma/files/7180279/Length-Experiment-Output.txt)

Attaching the output of experiment.  Printing default value in both cases.  EXPECTED DEFAULT VALUE : "0xffffff"

Comment (ulaskozat) [reply lte/gateway/python/scripts/generate_oai_config.py]: @panyogesh is orc8r connected to your set up? If service_mconfig.default_slice_differentiator is None, then len will fail.

Comment (ardzoht) [reply lte/gateway/python/scripts/generate_oai_config.py]: There are a couple of things to note here:

1. Currently there is no `defaultSliceDifferentiator` entry on the gateway.mconfig, so there won't be a parsed entry on the `service_mconfig` python object
2. Checking for `service_mconfig.default_slice_differentiator` existance should be enough, as Ulas says the call on `len()` will complain if there is no entry on there, which is this exact case.

So my comment still applies, I think we should just check for the entry existance `if service_mconfig.default_slice_differentiator is not None` to handle this entry. And it would be a good idea to include an entry for this on the `gateway.mconfig` default json file.

Comment (ardzoht) [reply lte/gateway/python/scripts/generate_oai_config.py]: And just to note, currently there is no Orc8r configs to build from for these entries, if the plan is to also add the configs to the cloud, `magma/lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml` should be updated to include these configs (either on gateway or network EPC configs?), this can be done on a separate PR but just wanted to bring the awareness for this. cc @ulaskozat

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 8]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    Args:
+        service_mconfig:
+
+    Returns: slice service type value
+    """
+
+    if len(service_mconfig.default_slice_service_type):
+        return service_mconfig.default_slice_service_type
+
+    return DEFAULT_NGAP_S_NSSAI_SST
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.default_slice_service_type or DEFAULT_NGAP_S_NSSAI_SST`

Comment (ulaskozat) [reply lte/gateway/python/scripts/generate_oai_config.py]: cc @ardzoht

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 9]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    it defaults to 0xffffff.
+    Args:
+        service_mconfig:
+
+    Returns: slice differentiator config value
+    """
+    if len(service_mconfig.default_slice_differentiator):
+        return service_mconfig.default_slice_differentiator
+
+    return DEFAULT_NGAP_S_NSSAI_SD
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.default_slice_differentiator or DEFAULT_NGAP_S_NSSAI_SD`

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 10]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    Args:
+        service_mconfig:
+
+    Returns: amf name string
+    """
+
+    if service_mconfig.amf_name is not None:
+        return service_mconfig.amf_name
+
+    return DEFAULT_NGAP_AMF_NAME
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.amf_name or DEFAULT_NGAP_AMF_NAME`

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 11]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    Args:
+        service_mconfig:
+
+    Returns: amf region id
+    """
+
+    if service_mconfig.amf_region_id is not None:
+        return service_mconfig.amf_region_id
+
+    return DEFAULT_NGAP_AMF_REGION_ID
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.amf_region_id or DEFAULT_NGAP_AMF_REGION_ID`

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 12]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    Args:
+        service_mconfig:
+
+    Returns: amf set id
+    """
+
+    if service_mconfig.amf_set_id is not None:
+        return service_mconfig.amf_set_id
+
+    return DEFAULT_NGAP_SET_ID
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.amf_set_id or DEFAULT_NGAP_SET_ID`

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 13]
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
...
+    Args:
+        service_mconfig:
+
+    Returns: amf pointer value
+    """
+
+    if service_mconfig.amf_pointer is not None:
+        return service_mconfig.amf_pointer
+
+    return DEFAULT_NGAP_AMF_POINTER
```
Comment (ulaskozat) [inline lte/gateway/python/scripts/generate_oai_config.py]: instead use `return service_mconfig.amf_pointer or DEFAULT_NGAP_AMF_POINTER`

File: lte/gateway/configs/gateway.mconfig [hunk 14]
```diff
@@ -39,7 +39,8 @@
       "ipv6DnsAddress": "2001:4860:4860:0:0:0:0:8888",
       "natEnabled": true,
       "congestionControlEnabled": true,
-      "enableConvergedCore": false
+      "enableConvergedCore": false,
```
Comment (ardzoht) [inline lte/gateway/configs/gateway.mconfig:42]: add entry for `defaultSliceDifferentiator` here as well

Comment (panyogesh) [reply lte/gateway/configs/gateway.mconfig:42]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 13 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 12 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 11 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 10 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 9 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

File: lte/gateway/python/scripts/generate_oai_config.py (hunk 8 above)
Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed.

**Comment by ulaskozat** [comment]
@panyogesh have you regenerated the pb files? insync-checkin is failing.

File: lte/gateway/configs/gateway.mconfig [hunk 15]
```diff
@@ -39,7 +39,9 @@
...
+      "defaultSliceServiceType": "1",
+      "defaultSliceDifferentiator" : "0x0"
```
Comment (ulaskozat) [inline lte/gateway/configs/gateway.mconfig]: why is this default different than the defaults defined in other parts?

**Comment by panyogesh** [comment]
Configured-SD-NGAP-Setup-Request-Sep-17
Default-NGAP-Setup-Request-Sep-17

File: lte/gateway/python/scripts/generate_oai_config.py [hunk 16]
```diff
@@ -38,6 +38,12 @@
 DEFAULT_DNS_IPV6_ADDR = "2001:4860:4860:0:0:0:0:8888"
 DEFAULT_P_CSCF_IPV4_ADDR = "172.27.23.150"
 DEFAULT_P_CSCF_IPV6_ADDR = "2a12:577:9941:f99c:0002:0001:c731:f114"
+DEFAULT_NGAP_S_NSSAI_SST = "1"
+DEFAULT_NGAP_S_NSSAI_SD = "0xffffff"
+DEFAULT_NGAP_AMF_NAME = "AMF_1"
+DEFAULT_NGAP_AMF_REGION_ID = "1"
+DEFAULT_NGAP_SET_ID = "1"
+DEFAULT_NGAP_AMF_POINTER = "1"
```
Comment (ardzoht) [inline lte/gateway/python/scripts/generate_oai_config.py]: this value for amf pointer is 1 but on the gateway.mconfig is set to 3, let's set them both to the expected default value. cc @ulaskozat

Comment (panyogesh) [reply lte/gateway/python/scripts/generate_oai_config.py]: Comments addressed. The default value of pointer should be 0.

File: lte/gateway/configs/gateway.mconfig (hunk 15 above)
Comment (panyogesh) [reply lte/gateway/configs/gateway.mconfig]: Comments addressed. Changed it to 0xffffff

**Comment by panyogesh** [comment]
ins

**Comment by panyogesh** [comment]
Default-NGAP-Setup-Request-Sep-17-1
